group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/ContextStateMachine.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.runtime;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.slf4j.Logger;
+
+/**
+ * ContextStateMachine is a wrapper on top of the coordinator state machine. Generally, only a single thread at a
+ * time will access this object but multiple threads may access while loading the __consumer_offsets topic partition.
```

Comment:
I think that we could do slightly better for the documentation. The object is indeed never accessed concurrently however it is accessed by multiple threads. Therefore, we need to sync'ed the access.

I would also remove the reference to the `__consumer_offsets` because the runtime is not tight to it. We can explain the same without mentioning it.